daemon/cmd/devices.go

```diff
@@ -0,0 +1,296 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2016-2021 Authors of Cilium
+
+// This module implements Cilium's network device detection.
+
+package cmd
+
+import (
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/datapath/linux/probes"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/mac"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+)
+
+var (
+	excludedDevicePrefixes = []string{
+		"cilium_",
+		"lo",
+		"lxc",
+		"cni",
+		"docker",
+	}
+)
+
+type DeviceManager struct {
+	mu      lock.Mutex
+	devices map[string]bool
+}
+
+func NewDeviceManager() *DeviceManager {
+	return &DeviceManager{
+		devices: make(map[string]bool),
+	}
+}
+
+func (dm *DeviceManager) Close() {
+}
+
+// Detect tries to detect devices which are to be used for:
+// NodePort BPF, direct routing in NodePort BPF, IPsec and Bandwidth Manager.
+//
+// The devices are detected by looking at all the configured global unicast
+// routes in the system.
+func (dm *DeviceManager) Detect() error {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+	dm.devices = make(map[string]bool)
+
+	if err := expandDevices(); err != nil {
+		log.WithError(err).Warnf("Failed to expand device wildcards")
+	}
+
+	l3DevOK := false
+	if option.Config.EnableNodePort {
+		// Probe whether fast redirect is supported for L3 devices. This will
+		// invoke bpftool and requires root privileges.
+		l3DevOK = supportL3Dev()
+	}
+
+	detectDevices := len(option.Config.Devices) == 0 &&
```

Comment:
If the device name expansion fails to find any matching devices, we will kick off the detection logic. Is that what we want? It may be a bit surprising for users if they make a mistake in the device pattern (not matching anything) and everything proceeds with unexpected devices (from detection).

Reply:
Good catch. I think it's best to fail if expansion fails completely. We do still want to allow expansion to partially not match so users can provide alternatives.